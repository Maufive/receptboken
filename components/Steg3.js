import React, { Component } from "react";
import PropTypes from "prop-types";
import { Loading } from "./Loading";
import { RecipeForm, DetailsContainer } from "./styles/Steg1Styles";
import { Button, DisabledButton } from "./styles/Button";
import SuccessIcon from "../svg/success.svg";

const TAGS = [
	"Texmex",
	"Vardagsmiddag",
	"Snabbt & Enkelt",
	"Italienskt",
	"Asiatiskt",
	"Smalt & Nyttigt",
	"Billigt",
	"Bakat & Dessert"
];

class Steg3 extends Component {
	state = {
		arr: [],
		photo: "",
		largePhoto: "",
		timeRequired: 0,
		tags: [],
		message: null,
		loading: false
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	handleTag = e => {
		const checkbox = e.target;
		// take a copy of the current permissions
		let updatedTags = [...this.state.tags];
		// figure out if we need to remove or add this permission
		if (checkbox.checked) {
			// add it in!
			updatedTags.push(checkbox.value);
		} else {
			updatedTags = updatedTags.filter(tags => tags !== checkbox.value);
		}
		this.setState({ tags: updatedTags });
	};

	uploadFile = async e => {
		this.setState({ loading: true });
		const files = e.target.files;
		const data = new FormData();
		data.append("file", files[0]);
		data.append("upload_preset", "receptboken");

		const res = await fetch(
			"https://api.cloudinary.com/v1_1/dcgb0fhog/image/upload",
			{
				method: "POST",
				body: data
			}
		).catch(error => this.props.setMessage("danger", error));
		const file = await res.json();
		this.setState({
			photo: file.secure_url,
			largePhoto: file.eager[0].secure_url,
			loading: false
		});
	};

	saveState = async e => {
		e.preventDefault();
		await this.props.saveDetails(
			this.state.photo,
			this.state.timeRequired,
			this.state.tags
		);
		this.props.submitRecipe();
	};

	render() {
		const { photo, timeRequired } = this.state;
		return (
			<RecipeForm>
				<h3>{this.props.steg}. Fyll i lite detaljer om ditt recept</h3>

				<DetailsContainer>
					<div>
						<label htmlFor="timeRequired">
							Ungefär hur lång tid tar receptet?{" "}
						</label>
						<div>
							<input
								type="range"
								name="timeRequired"
								min={0}
								max={180}
								value={timeRequired}
								onChange={this.saveToState}
							/>
							<p>{timeRequired} minuter</p>
						</div>
					</div>
					<div>
						<label htmlFor="photo">Ladda upp en bild:</label> <br />
						<input
							type="file"
							id="file"
							name="photo"
							placeholder="Ladda upp en bild"
							onChange={this.uploadFile}
						/>
						<br />
						{this.state.loading && <Loading />}
						{photo && (
							<img width="300" src={photo} alt="Bild förhandsvisning" />
						)}
					</div>
					<div>
						<label htmlFor="tags">Kryssa i passande kategorier: </label> <br />
						{TAGS.map(tag => (
							<label htmlFor={`recipe-tag-${tag}`} key={tag}>
								<input
									type="checkbox"
									id={`recipe-tag-${tag}`}
									checked={this.state.tags.includes(tag)}
									value={tag}
									onChange={this.handleTag}
								/>{" "}
								{tag} <br />
							</label>
						))}
					</div>
				</DetailsContainer>

				<div style={{ display: "flex" }}>
					<Button fullWidth onClick={this.props.previousStep}>
						← Föregående
					</Button>
					{this.state.timeRequired <= 4 && (
						<DisabledButton fullWidth>
							<SuccessIcon /> Ladda upp!
						</DisabledButton>
					)}
					{this.state.timeRequired >= 5 && (
						<Button primary fullWidth onClick={this.saveState}>
							<SuccessIcon /> Ladda upp!
						</Button>
					)}
				</div>
			</RecipeForm>
		);
	}
}

export default Steg3;

Steg3.propTypes = {
	steg: PropTypes.number,
	previousStep: PropTypes.func,
	saveDetails: PropTypes.func,
	setMessage: PropTypes.func
};
