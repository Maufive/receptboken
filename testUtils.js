export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeShoppinglist = [
	{
		_id: "34567876543452176",
		title: "My Shoppinglist",
		created: "12/3 2019",
		author: "61386712376576d7q6ewd65",
		list: [
			{
				input: "Lök",
				numberOfUnits: "5",
				units: "dl"
			},
			{
				input: "Tomater",
				numberOfUnits: "10",
				units: "st"
			}
		]
	}
];
