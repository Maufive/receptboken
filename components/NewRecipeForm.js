import React, { Component } from "react";
import { Header, RecipeForm } from "./styles/Steg1Styles";
import Steg1 from "./Steg1";
import Steg2 from "./Steg2";
import Steg3 from "./Steg3";

class NewRecipeForm extends Component {
	state = {
		title: "",
		description: [],
		tags: [],
		timeRequired: 0,
		photo: "",
		ingredients: [],
		servings: 2,
		step: 1
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	previousStep = () => {
		this.setState({ step: this.state.step - 1 });
	};

	saveIngredients = (arr, servings) => {
		this.setState({
			ingredients: arr,
			servings,
			step: this.state.step + 1
		});
	};

	saveDescription = arr => {
		this.setState({
			description: arr,
			step: this.state.step + 1
		});
	};

	saveDetails = (photo, time, tags) => {
		this.setState({
			photo,
			timeRequired: time,
			tags,
			step: this.state.step + 1
		});
	};

	render() {
		const step = this.state.step;
		let ActiveComponent;
		// För varje steg i processen att skapa ett nytt recept så renderar jag en specifik component.
		// Syftet är att dela upp en lång process att ladda upp ett recept i mindre bitar för att göra det
		// lättare för användare.
		if (step === 1) {
			ActiveComponent = (
				<Steg1
					saveToState={this.saveToState}
					saveIngredients={this.saveIngredients}
					ingredients={this.state.ingredients}
					servings={this.state.servings}
				/>
			);
		} else if (step === 2) {
			ActiveComponent = (
				<Steg2
					saveDescription={this.saveDescription}
					steg={step}
					previousStep={this.previousStep}
					description={this.state.description}
				/>
			);
		} else if (step === 3) {
			ActiveComponent = (
				<Steg3
					steg={step}
					previousStep={this.previousStep}
					saveDetails={this.saveDetails}
				/>
			);
		} else {
			ActiveComponent = <h2>Finished!</h2>;
			// To-do: Skriv en funktion för en POST-method att passa ner till steg 3. Spara Allt som finns i state här
			// till databasen.
			// Gör en LOADING komponent som snurrar medans man väntar på att databasen ska spara receptet
			// Gör en Page för ett recept där jag kan displaya informationen
			// Redirecta användaren när man lyckats spara ett recept till databasen till det receptet

		}

		return (
			<div>
				<Header>
					<input
						type="text"
						value={this.state.title}
						onChange={this.saveToState}
						name="title"
						required
						placeholder="Nytt recept..."
					/>
					<i onClick={this.editTitle} className="icofont-ui-edit" />
				</Header>
				<RecipeForm>{ActiveComponent}</RecipeForm>
			</div>
		);
	}
}

export default NewRecipeForm;
