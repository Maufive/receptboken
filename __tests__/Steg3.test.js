import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Steg3 from "../components/Steg3";
import NewRecipeForm from "../components/NewRecipeForm";
import { fakeUser } from "../testUtils";

describe("<Steg3 />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<Steg3 />)));

	it("renders without error and matches snapshot", () => {
		expect(wrapper.find("Steg1Styles__RecipeForm").exists()).toBe(true);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	describe("Time required input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#timeRequired").length).toEqual(1);
		});

		it("handles the onChange event", () => {
			wrapper.find("#timeRequired").simulate("change", {
				target: { name: "timeRequired", value: 60 }
			});
			expect(wrapper.state().timeRequired).toEqual(60);
		});
	});

	describe("File input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#file").length).toEqual(1);
		});
	});
});
