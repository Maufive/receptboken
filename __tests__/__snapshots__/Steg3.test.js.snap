// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steg3 /> renders without error and matches snapshot 1`] = `
<Steg1Styles__RecipeForm>
  <h3
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <span>
      Ingedienser
    </span>
    <span>
      Steg 
       av 3
    </span>
  </h3>
  <Steg1Styles__DetailsContainer>
    <div>
      <label
        htmlFor="timeRequired"
      >
        Ungefär hur lång tid tar receptet?
         
      </label>
      <div>
        <input
          id="timeRequired"
          max={180}
          min={0}
          name="timeRequired"
          onChange={[Function]}
          type="range"
          value={0}
        />
        <p>
          0
           minuter
        </p>
      </div>
    </div>
    <div>
      <label
        htmlFor="photo"
      >
        Ladda upp en bild:
      </label>
       
      <br />
      <input
        id="file"
        name="photo"
        onChange={[Function]}
        placeholder="Ladda upp en bild"
        type="file"
      />
      <br />
    </div>
    <div>
      <label
        htmlFor="tags"
      >
        Kryssa i passande kategorier: 
      </label>
       
      <br />
      <label
        htmlFor="recipe-tag-Texmex"
        key="Texmex"
      >
        <input
          checked={false}
          id="recipe-tag-Texmex"
          onChange={[Function]}
          type="checkbox"
          value="Texmex"
        />
         
        Texmex
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Vardagsmiddag"
        key="Vardagsmiddag"
      >
        <input
          checked={false}
          id="recipe-tag-Vardagsmiddag"
          onChange={[Function]}
          type="checkbox"
          value="Vardagsmiddag"
        />
         
        Vardagsmiddag
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Snabbt & Enkelt"
        key="Snabbt & Enkelt"
      >
        <input
          checked={false}
          id="recipe-tag-Snabbt & Enkelt"
          onChange={[Function]}
          type="checkbox"
          value="Snabbt & Enkelt"
        />
         
        Snabbt & Enkelt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Italienskt"
        key="Italienskt"
      >
        <input
          checked={false}
          id="recipe-tag-Italienskt"
          onChange={[Function]}
          type="checkbox"
          value="Italienskt"
        />
         
        Italienskt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Asiatiskt"
        key="Asiatiskt"
      >
        <input
          checked={false}
          id="recipe-tag-Asiatiskt"
          onChange={[Function]}
          type="checkbox"
          value="Asiatiskt"
        />
         
        Asiatiskt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Smalt & Nyttigt"
        key="Smalt & Nyttigt"
      >
        <input
          checked={false}
          id="recipe-tag-Smalt & Nyttigt"
          onChange={[Function]}
          type="checkbox"
          value="Smalt & Nyttigt"
        />
         
        Smalt & Nyttigt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Billigt"
        key="Billigt"
      >
        <input
          checked={false}
          id="recipe-tag-Billigt"
          onChange={[Function]}
          type="checkbox"
          value="Billigt"
        />
         
        Billigt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Bakat & Dessert"
        key="Bakat & Dessert"
      >
        <input
          checked={false}
          id="recipe-tag-Bakat & Dessert"
          onChange={[Function]}
          type="checkbox"
          value="Bakat & Dessert"
        />
         
        Bakat & Dessert
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Pasta"
        key="Pasta"
      >
        <input
          checked={false}
          id="recipe-tag-Pasta"
          onChange={[Function]}
          type="checkbox"
          value="Pasta"
        />
         
        Pasta
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Vegetariskt"
        key="Vegetariskt"
      >
        <input
          checked={false}
          id="recipe-tag-Vegetariskt"
          onChange={[Function]}
          type="checkbox"
          value="Vegetariskt"
        />
         
        Vegetariskt
         
        <br />
      </label>
      <label
        htmlFor="recipe-tag-Fisk"
        key="Fisk"
      >
        <input
          checked={false}
          id="recipe-tag-Fisk"
          onChange={[Function]}
          type="checkbox"
          value="Fisk"
        />
         
        Fisk
         
        <br />
      </label>
    </div>
  </Steg1Styles__DetailsContainer>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Button
      fullWidth={true}
    >
      ← Föregående
    </Button>
    <Button__DisabledButton
      fullWidth={true}
    >
      <SuccessIcon
        viewBox="0 0 52 52"
        xmlns="http://www.w3.org/2000/svg"
      />
       Ladda upp!
    </Button__DisabledButton>
  </div>
</Steg1Styles__RecipeForm>
`;
